// BatHat/node_modules/ModelBase.js
var _ = require('underscore');
var FieldBase = require('FieldBase');

// -----------------------------------------------------------------------------
// When you define a model, Frank takes care of setting up the MongoDB
// Collection.
// 
// It also provides a nice 'ORM' to your models, so you don't need to type
// MongoDB queries all the time.
// -----------------------------------------------------------------------------
var ModelBase = function(options)
{
	this.fields = {};
	this.scopes = {};
	this.db = null;

	if(options['timestamps'])
	{
		this.fields['created_at'] =
		{
			'type': 'datetime',
			'default': Date.now()
		};

		this.fields['updated_at'] = 
		{
			'type': 'datetime',
			'default': Date.now()
		};
	}

	if(options['paranoia'])
	{
		this.fields['deleted_at'] = 
		{
			'type': 'datetime',
			'default': Date.now()
		};
	}

	_.extend(this.fields, options.fields);
	_.extend(this, options.scopes);
}

ModelBase.prototype = 
{
	// -------------------------------------------------------------------------
	// It takes the model as you setup to work with the database,
	// and converts it to a fully functional Backbone Model. 
	// 
	// Returns a String that you need to send to the client browser
	// -------------------------------------------------------------------------
	toBackbone: function()
	{
	},

	// -------------------------------------------------------------------------
	// Querying interface
	// -------------------------------------------------------------------------
	first: function()
	{
	},

	last: function()
	{
	},

	all: function(db)
	{
		db.collection('users', function(err, collection)
		{
			collection.find().toArray(function(err, items)
			{
				console.log(items);
			});
		});
	},

	find: function(id)
	{
	}

	// -------------------------------------------------------------------------
	// Easy validation
	// -------------------------------------------------------------------------
}


// -----------------------------------------------------------------------------
// Usage:
// var User = Frant.ModelBase.extend({options});
module.exports = 
{
	extend: function(options)
	{
		return new ModelBase(options);
	}
}
