// BatHat/node_modules/ModelBase.js
var _ = require('underscore');
var FieldBase = require('FieldBase');
var ObjectID = require('mongodb').ObjectID;

// -----------------------------------------------------------------------------
// When you define a model, Frank takes care of setting up the MongoDB
// Collection.
// 
// It also provides a nice 'ORM' to your models, so you don't need to type
// MongoDB queries all the time.
// -----------------------------------------------------------------------------
var ModelBase = function(options)
{
	this.fields = {};
	this.scopes = {};
	this.db = null;
	this.name = null;
	this.FieldBase = require('FieldBase');

	if(options['timestamps'])
	{
		this.fields['created_at'] =
		{
			'type': 'datetime',
			'default': Date.now()
		};

		this.fields['updated_at'] = 
		{
			'type': 'datetime',
			'default': Date.now()
		};
	}

	if(options['paranoia'])
	{
		this.fields['deleted_at'] = 
		{
			'type': 'datetime',
			'default': Date.now()
		};
	}

	_.extend(this.fields, options.fields);
	_.extend(this, options.scopes);
}

ModelBase.prototype = 
{
	// -------------------------------------------------------------------------
	// It takes the model as you setup to work with the database,
	// and converts it to a fully functional Backbone Model. 
	// 
	// Returns a String that you need to send to the client browser
	// -------------------------------------------------------------------------
	toBackbone: function()
	{
	},

	setDb: function(db)
	{
		this.db = db;
	},

	// -------------------------------------------------------------------------
	// Querying interface
	// -------------------------------------------------------------------------

	// -------------------------------------------------------------------------
	// FIRST
	// -------------------------------------------------------------------------
	first: function(callback)
	{
		if(this.db === null) return;
		var root = this;
		this.db.collection(this.name, function(err, collection)
		{
			collection.find().sort({_id: 1}).limit(1).toArray
			(
				function(err, item)
				{
					callback(err, item[0]);
				}
			);
		});
	},
	firstWrapped: function(callback)
	{
		if(this.db === null) return;
		var root = this;
		this.db.collection(this.name, function(err, collection)
		{
			collection.find().sort({_id: 1}).limit(1).toArray
			(
				function(err, item)
				{
					callback(err, root.FieldBase.create(root.db, root.name,
						item));
				}
			);
		});
	},

	// -------------------------------------------------------------------------
	// LAST
	// -------------------------------------------------------------------------
	last: function()
	{
	},

	// -------------------------------------------------------------------------
	// ALL
	// -------------------------------------------------------------------------
	all: function(callback)
	{
		if(this.db === null) return;
		var root = this;
		this.db.collection(this.name, function(err, collection)
		{
			collection.find().toArray(function(err, items)
			{
				callback(err, items);
			});
		});
	},
	allWrapped: function(callback)
	{
		if(this.db === null) return;
		var root = this;
		this.db.collection(this.name, function(err, collection)
		{
			collection.find().toArray(function(err, items)
			{
				var tItems = [];
				for(var i = 0; i < items.length; i++)
				{
					tItems[i] = new root.FieldBase(root.db, root.name, 
						items[i]);
				}
				callback(err, tItems);
			});
		});
	},

	find: function(id, callback)
	{
		if(this.db === null) return;
		var root = this;
		this.db.collection(this.name, function(err, collection)
		{
			collection.findOne(
				{ _id: new ObjectID(id) }, 
				function(err, item)
				{
					callback(item);
				});
		});
	}

	// -------------------------------------------------------------------------
	// Easy validation
	// -------------------------------------------------------------------------
}


// -----------------------------------------------------------------------------
// Usage:
// var User = Frant.ModelBase.extend({options});
module.exports = 
{
	extend: function(options)
	{
		return new ModelBase(options);
	}
}
